minidump-processor/src/processor.rs

```diff
@@ -763,3 +769,103 @@ where
+/// an address contain the same binary digit as bit 48. A violation of this rule triggers a
+/// General Protection Fault instead of the usual Page Fault, which unfortunately means that the
+/// OS has no idea what memory address actually caused the issue
+///
+/// If `exception_info` contains the markers of a non-canonical exception, and it also contains
+/// memory access info from analyzing the CPU instruction with `op_analysis`, this module will
+/// attempt to determine what address the CPU was instructed to access when the GPF occurred
+///
+/// Nothing will be changed if the crash was not caused by a non-canonical access
+fn fix_non_canonical_crash_address(
+    system_info: &SystemInfo,
+    exception_info: &mut crate::ExceptionInfo,
+) {
+    use minidump_common::errors as minidump_errors;
+    use system_info::{Cpu, Os};
+
+    // This is needed because match arms don't allow casting
+    const SI_KERNEL_U32: u32 = minidump_errors::ExceptionCodeLinuxSicode::SI_KERNEL as u32;
+
+    // The range of non-canonical addresses in the current 48-bit implementation
+    // See: https://en.wikipedia.org/wiki/X86-64#Virtual_address_space_details
+    const NON_CANONICAL_RANGE: RangeInclusive<u64> = 0x0000_8000_0000_0000..=0xffff_7fff_ffff_ffff;
+
+    // Only Amd64 has non-canonical addresses
+    if system_info.cpu != Cpu::X86_64 {
+        return;
+    }
+
+    // Different OSes have unique ways of reporting this error
+    let is_non_canonical_access = match (
+        system_info.os,
+        exception_info.reason,
+        exception_info.address,
+    ) {
+        // Windows reports it as EXCEPTION_ACCESS_VIOLATION_READ, address 0xffffffffffffffff
+        (
+            Os::Windows,
+            CrashReason::WindowsAccessViolation(
+                minidump_errors::ExceptionCodeWindowsAccessType::READ,
+            ),
+            u64::MAX,
+        ) => true,
+        (Os::Windows, _, _) => false,
+        // macOS reports it as EXC_BAD_ACCESS / EXC_I386_GPFLT, address 0x0000000000000000
+        (
+            Os::MacOs,
+            CrashReason::MacBadAccessX86(
+                minidump_errors::ExceptionCodeMacBadAccessX86Type::EXC_I386_GPFLT,
+            ),
+            0,
+        ) => true,
+        (Os::MacOs, _, _) => false,
+        // Linux reports it as either "SIGBUS / SI_KERNEL" or "SIGSEGV / SI_KERNEL", address 0x0000000000000000
+        (
+            Os::Linux,
+            CrashReason::LinuxGeneral(minidump_errors::ExceptionCodeLinux::SIGSEGV, SI_KERNEL_U32),
+            0,
+        ) => true,
+        (
+            Os::Linux,
+            CrashReason::LinuxGeneral(minidump_errors::ExceptionCodeLinux::SIGBUS, SI_KERNEL_U32),
+            0,
+        ) => true,
+        (Os::Linux, _, _) => false,
+        (_, _, _) => {
+            tracing::warn!("we don't currently support non-canonical analysis for your OS");
+            return;
+        }
+    };
```

Comment:
It would be nice to wrap this large check into its own function so you could do something like:
```rust
if !is_non_canonical_access(...) {
    return;
}
```
below

Reply:
Works for me! Done :)